lib/Wikia/src/Consul/Client.php

```diff
@@ -57,31 +60,128 @@ function( $item ) {
 	}
 
 	/**
+	 * Returns IP addresses (with ports) of given service healthy nodes
+	 *
+	 * $catalog->getNodes( 'db-a', 'slave' )
+	 * $catalog->getNodes( 'chat-private', 'prod' )
+	 *
+	 * $ curl 'http://127.0.0.1:8500/v1/query/geo-db-a-slave/execute?pretty=&passing='
+	 *
+	 * @param string $query
+	 * @return array list of IP addresses with ports ie. 127.0.0.1:1234
+	 */
+	private function getNodesFromConsulQuery( string $query ) : array {
+		$consulClient = new ConsulClient();
+		$resp = $consulClient->get(
+			sprintf( '/v1/query/%s/execute?passing=', $query )
+		)->json();
+
+		$nodes = array_map(
+			function( $item ) {
+				return $item[ 'Node' ][ 'Address' ] . ':' . $item[ 'Service' ][ 'Port' ];
+			},
+			$resp['Nodes']
+		);
+
+		wfDebug( __METHOD__ .sprintf( ": got nodes for '%s' query: %s\n", $query, join(', ', $nodes) ) );
+		return $nodes;
+	}
+
+	/**
 	 * Helper method for getting IP addresses of all nodes hidden behind consul
 	 *
-	 * $catalog->getNodesFromHostname( 'slave.db-smw.service.consul' )
+	 * $consul->getNodesFromHostname( 'slave.db-smw.service.consul' )
+	 * $consul->getNodesFromHostname( 'geo-db-g-slave.query.consul' )
 	 *
 	 * @param string $hostname
 	 * @return array list of IP addresses
+	 * @throws \Exception
 	 */
-	function getNodesFromHostname( $hostname ) {
+	function getNodesFromHostname( string $hostname ) {
 		Assert::true( self::isConsulAddress( $hostname ), __METHOD__ . ' should get a Consul address', $this->getLoggerContext() );
 
-		list( $tag, $service ) = explode( '.', $hostname );
-		return $this->getNodes( $service, $tag );
+		if ( self::isConsulQuery($hostname) ) {
+			// e.g. geo-db-g-slave.query.consul
+			$query = self::parseConsulQuery( $hostname );
+			return $this->getNodesFromConsulQuery( $query );
+		}
+		elseif ( self::isConsulServiceAddress( $hostname ) ) {
+			// e.g. slave.db-g.service.consul
+			list( $tag, $service ) = self::parseConsulServiceAddress( $hostname);
+			return $this->getNodes( $service, $tag );
+		}
+		else {
+			throw new \Exception( __METHOD__ . " - {$hostname} is neither consul query nor consul service address" );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)

[project-id]: wikia:app